src/external/ocaml-rocksdb/rocks.ml

```diff
@@ -690,6 +690,61 @@ and RocksDb : (Rocks_intf.ROCKS with type batch := WriteBatch.t) = struct
     | None ->
         ReadOptions.with_t inner
 
+  let multi_get_raw =
+    foreign "rocksdb_multi_get"
+      ( t
+      @-> ReadOptions.t
+      @-> Views.int_to_size_t
+      @-> ptr (ptr char)
+      @-> ptr size_t
+      @-> ptr (ptr char)
+      @-> ptr size_t
+      @-> ptr (ptr char)
+      @-> returning void )
+
+  let multi_get ?opts t keys =
+    let inner opts =
+      let cast_array_ptr typ arr = arr |> bigarray_start array1 |> to_voidp |> from_voidp typ in
+      let count = List.length keys in
+      (* could optimize this into a single allocation *)
+      let inputs = Bigarray.(Array1.create Nativeint C_layout) count in
+      let input_lengths = Bigarray.(Array1.create Nativeint C_layout) count in
+      let outputs = Bigarray.(Array1.create Nativeint C_layout) count in
+      let output_lengths = Bigarray.(Array1.create Nativeint C_layout) count in
+      let errors = Bigarray.(Array1.create Nativeint C_layout) count in
+      keys |> List.iteri (fun i key ->
```

Comment:
It looks like this is the last time `keys` is referred to in this function. Doesn't this mean that some of the earlier keys in the list may be freed while allocating the `bigarray_start`s for later `key`s? To be safe, I think this needs an `ignore keys` after the `get_multi_raw` call to ensure that the bigstrings outlive that call.

Reply:
Good catch. I will use Janestreet's `keep_alive` trick instead of `ignore` so that the compiler doesn't optimize this guarantee away. https://github.com/janestreet/core_kernel/blob/master/core/src/gc.ml#L354